src/Workspaces/Remote/Core/AbstractAssetProvider.cs

```diff
@@ -72,13 +72,27 @@ public async Task<SolutionInfo> CreateSolutionInfoAsync(Checksum solutionChecksu
             additionalDocumentInfos,
             analyzerConfigDocumentInfos,
             hostObjectType: null); // TODO: https://github.com/dotnet/roslyn/issues/62804
+
+        async Task<ImmutableArray<DocumentInfo>> CreateDocumentInfosAsync(ChecksumCollection documentChecksums)
+        {
+            using var _ = ArrayBuilder<DocumentInfo>.GetInstance(documentChecksums.Count, out var documentInfos);
+
+            foreach (var documentChecksum in documentChecksums)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                documentInfos.Add(await CreateDocumentInfoAsync(projectId, documentChecksum, cancellationToken).ConfigureAwait(false));
+            }
+
+            return documentInfos.ToImmutableAndClear();
+        }
     }
 
-    public async Task<DocumentInfo> CreateDocumentInfoAsync(Checksum documentChecksum, CancellationToken cancellationToken)
+    public async Task<DocumentInfo> CreateDocumentInfoAsync(
+        ProjectId projectId, Checksum documentChecksum, CancellationToken cancellationToken)
     {
-        var documentSnapshot = await GetAssetAsync<DocumentStateChecksums>(documentChecksum, cancellationToken).ConfigureAwait(false);
-        var attributes = await GetAssetAsync<DocumentInfo.DocumentAttributes>(documentSnapshot.Info, cancellationToken).ConfigureAwait(false);
-        var serializableSourceText = await GetAssetAsync<SerializableSourceText>(documentSnapshot.Text, cancellationToken).ConfigureAwait(false);
+        var documentSnapshot = await GetAssetAsync<DocumentStateChecksums>(hintProject: projectId, documentChecksum, cancellationToken).ConfigureAwait(false);
+        var attributes = await GetAssetAsync<DocumentInfo.DocumentAttributes>(hintProject: projectId, documentSnapshot.Info, cancellationToken).ConfigureAwait(false);
+        var serializableSourceText = await GetAssetAsync<SerializableSourceText>(hintProject: projectId, documentSnapshot.Text, cancellationToken).ConfigureAwait(false);
```

Comment:
this scopes us down to a particular project to find this data.  but we still need to search the whole project to find the document info being asked for.  a followup PR addresses this by allowing the doc-id to be passed along as well to narrow the search even further.